### ISSUE TITLE
[lambda]Incorrect Lambda function version calculation causes deployment difficulty

### ISSUE THREAD

**Issue by aidansteele** [aws/aws-cdk#11537]
This is :bug: Bug Report.

It is possible to attempt get an error from CloudFormation when making a legitimate change to a Lambda function in a CDK stack.

### Reproduction Steps

Here is an example stack:

```typescript
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';

export class SampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new lambda.Function(this, 'Function', {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.InlineCode.fromInline(`exports.handler = (event) => "hello world"`),
      // reservedConcurrentExecutions: 10 // after creating stack, uncomment this line and re-deploy
    }).currentVersion.addAlias('live');
  }
}
```

### What did you expect to happen?

When I uncomment the `reservedConcurrentExecutions: 10` property, I expect `cdk deploy` to update the Lambda function's reserved concurrency configuration.

### What actually happened?

I get the following error from the CloudFormation service:

```
% npx cdk deploy
SampleStack: deploying...
SampleStack: creating CloudFormation changeset...
3:59:04 PM | CREATE_FAILED        | AWS::Lambda::Version  | FunctionCurrentVer...faacc45be2b3a1451e
A version for this Lambda function exists ( 1 ). Modify the function to create a new version.
```

### Environment

  - **CDK CLI Version:** `1.74.0 (build e86602f)`
  - **Node.js Version:** v15.0.1 
  - **OS               :** MacOS
  - **Language (Version):** all

### Suspected cause

I suspect the responsible code is these lines:

https://github.com/aws/aws-cdk/blob/33cc5ec63d648e3d93a51e78d396b6aee16e37ea/packages/%40aws-cdk/aws-lambda/lib/function-hash.ts#L11-L14

The issue is that the string generated by `JSON.stringify(config)` on line 14 differs like this between the two iterations of the stack (prettified):

```diff
--- before	2020-11-18 16:17:45.000000000 +1100
+++ after	2020-11-18 16:17:37.000000000 +1100
@@ -1,26 +1,27 @@
 {
   "Resources": {
     "Function76856677": {
       "Type": "AWS::Lambda::Function",
       "Properties": {
         "Code": {
           "ZipFile": "exports.handler = (event) => \"hello world\""
         },
         "Handler": "index.handler",
         "Role": {
           "Fn::GetAtt": [
             "FunctionServiceRole675BB04A",
             "Arn"
           ]
         },
-        "Runtime": "nodejs12.x"
+        "Runtime": "nodejs12.x",
+        "ReservedConcurrentExecutions": 10
       },
       "DependsOn": [
         "FunctionServiceRole675BB04A"
       ],
       "Metadata": {
         "aws:cdk:path": "SampleStack/Function/Resource"
       }
     }
   }
 }
\ No newline at end of file

```

This causes the `AWS::Lambda::Version` resource to have a different logical ID:

```diff
--- before	2020-11-18 16:15:54.000000000 +1100
+++ after	2020-11-18 16:16:07.000000000 +1100
@@ -40,49 +40,50 @@
         "Code": {
           "ZipFile": "exports.handler = (event) => \"hello world\""
         },
         "Handler": "index.handler",
         "Role": {
           "Fn::GetAtt": [
             "FunctionServiceRole675BB04A",
             "Arn"
           ]
         },
-        "Runtime": "nodejs12.x"
+        "Runtime": "nodejs12.x",
+        "ReservedConcurrentExecutions": 10
       },
       "DependsOn": [
         "FunctionServiceRole675BB04A"
       ],
       "Metadata": {
         "aws:cdk:path": "SampleStack/Function/Resource"
       }
     },
-    "FunctionCurrentVersion4E2B226126dc1a973a7e75ab5d120224c6c45fb3": {
+    "FunctionCurrentVersion4E2B2261de715023e30143628ce2452cd16d155d": {
       "Type": "AWS::Lambda::Version",
       "Properties": {
         "FunctionName": {
           "Ref": "Function76856677"
         }
       },
       "Metadata": {
         "aws:cdk:path": "SampleStack/Function/CurrentVersion/Resource"
       }
     },
     "FunctionCurrentVersionAliaslive2D187DC4": {
       "Type": "AWS::Lambda::Alias",
       "Properties": {
         "FunctionName": {
           "Ref": "Function76856677"
         },
         "FunctionVersion": {
           "Fn::GetAtt": [
-            "FunctionCurrentVersion4E2B226126dc1a973a7e75ab5d120224c6c45fb3",
+            "FunctionCurrentVersion4E2B2261de715023e30143628ce2452cd16d155d",
             "Version"
           ]
         },
         "Name": "live"
       },
       "Metadata": {
         "aws:cdk:path": "SampleStack/Function/CurrentVersion/Aliaslive/Resource"
       }
     },
     "CDKMetadata": {

```

Which in turn causes CloudFormation to return that error. **This problem occurs whenever the CloudFormation function resource synthesised by CDK differs in _any_ fields that the Lambda service does not include in its versioning.** Such fields include e.g. `Properties.ReservedConcurrentExecutions`, `Properties.Tags`, and `DependsOn`. We should instead calculate a hash of only the [versionable properties of a function](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionConfiguration.html). 

Let me know if you need any more details or anything else from me!

**Comment by nija-at** [comment]
_New report filed by @jmm at #14428_

Hello,

With CDK 1.101.0, I'm encountering this when attempting to deploy an updated stack with an unchanged Lambda, but where I've changed the policy configuration of its execution role, e.g. with `someFunction.role.addToPrincipalPolicy()`.

Somewhat related to https://github.com/aws/aws-cdk/issues/5334.

So in this case the Lambda resource in the CloudFormation template is identical except for the addition of an AWS::IAM::Policy's name to its `DependsOn` set.

Based on the description of https://github.com/aws/aws-cdk/pull/6771 and a skim of the code, I'm assuming it's incorporating the function's `DependsOn` attribute in the hashing. And based on the documentation of `DependsOn`, it sounds like that should be irrelevant for the purpose of https://github.com/aws/aws-cdk/pull/6771, right? What do you think about omitting it from the hashing?

So in plain CloudFormation it doesn't automatically create a new version in response to changes to the function code or configuration, correct? You either need to manually change something about the version configuration, like description, or wire up something to detect changes to the function yourself and do something that forces CloudFormation to create a new version, like here you're changing the logical ID.

I'm temporarily using the workaround from https://github.com/aws/aws-cdk/issues/5334.

### Reproduction Steps

First iteration:

```js
const someFunction = new lambda.Function(this, "some-function", {
  code: lambda.Code.fromAsset(somePath),
  handler: "index.lambdaHandler",
  runtime: lambda.Runtime.NODEJS_14_X,
})
```

Second iteration:

```js
const someFunction = new lambda.Function(this, "some-function", {
  code: lambda.Code.fromAsset(somePath),
  handler: "index.lambdaHandler",
  runtime: lambda.Runtime.NODEJS_14_X,
})

someFunction.role.addToPrincipalPolicy(
  new iam.PolicyStatement({
    actions: ["s3:GetObject"],
    effect: iam.Effect.ALLOW,
    resources: [bucketArn],
  })
)
```

### What did you expect to happen?

Both iterations to deploy successfully, modifying the Lambda's execution role in the second one.

### What actually happened?

On the second iteration, failed deployment with error message (where `{n}` is a placeholder):

> A version for this Lambda function exists ( {n} ). Modify the function to create a new version.

**Comment by alexandervandekleutab** [comment]
I've also run into this issue adding to the lambda's role policy after creation. I will try deleting and re-creating for now.

**Comment by MaRuifeng** [comment]
@nija-at Any chance we can revisit/reopen this issue? It's recurring in CDK v2.

**Comment by kornicameister** [comment]
happened for me as well, function was not changed at all though. 
I had to fix `Fn::Export/Fn::Import` clash and deployed stack of versioned lambda first. Second deployment tried to deploy lambda-stack as a result of a dependency. Once again, I haven't changed a bit in function configuration.

**Comment by bes** [comment]
Happened to me just now when upgrading from CDK v1 to CDK v2. The "Change description hack" worked.

**Comment by anhtumai-spacehub** [comment]
has this error got fixed? and in which version?

**Comment by jerry42** [comment]
FYI : Just got the problem (cdk v2.59.0) and used the workaround https://github.com/aws/aws-cdk/issues/5334

Thanks !

**Comment by mhvelplund** [comment]
> FYI : Just got the problem (cdk v2.59.0) and used the workaround #5334

TL;DR: the workaround is change the description: https://github.com/aws/aws-cdk/issues/5334#issuecomment-562981777
